transaction-multi-payment/src/lib.rs

```diff
@@ -361,68 +400,120 @@ where
+        }
+        let fee_receiver = DP::get_fee_receiver().ok_or(TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+
+        // get the currency in which fees are paid. In case of non-native currency, the price is required to calculate final fee.
+        let maybe_non_native = DP::get_currency_and_price(who).map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+        match maybe_non_native {
+            (_, None) => {
+                // match C::withdraw(who, fee, withdraw_reason, ExistenceRequirement::KeepAlive) {
+                match C::transfer(who, &fee_receiver, fee, ExistenceRequirement::KeepAlive) {
+                    Ok(()) => Ok(Some(PaymentInfo::Native(fee))),
+                    Err(_) => Err(InvalidTransaction::Payment.into()),
+                }
+            },
+            (currency, Some(price)) => {
+                let converted_fee = price.checked_mul_int(fee).ok_or(TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+                match MC::transfer(currency.into(), who, &fee_receiver, converted_fee.into()) {
+                    Ok(()) => Ok(Some(PaymentInfo::NonNative(converted_fee, currency, price))),
+                    Err(_) => Err(InvalidTransaction::Payment.into()),
+                }
+            },
+        }
+    }
+
+    /// Since the predicted fee might have been too high, parts of the fee may
+    /// be refunded.
+    ///
+    /// Note: The `fee` already includes the `tip`.
+    fn correct_and_deposit_fee(
+        who: &T::AccountId,
+        _dispatch_info: &DispatchInfoOf<T::Call>,
+        _post_info: &PostDispatchInfoOf<T::Call>,
+        corrected_fee: Self::Balance,
+        _tip: Self::Balance,
+        already_withdrawn: Self::LiquidityInfo,
+    ) -> Result<(), TransactionValidityError> {
+        let fee_receiver = DP::get_fee_receiver().ok_or(TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+
+        if let Some(paid) = already_withdrawn {
+            // Calculate how much refund we should return and transfer the amount to the account that paid the fees
+            match paid {
+                PaymentInfo::Native(paid_fee) => {
+                    let refund_amount = paid_fee.saturating_sub(corrected_fee);
+                    C::transfer(&fee_receiver, who, refund_amount, ExistenceRequirement::KeepAlive)
+                        .map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+                },
+                PaymentInfo::NonNative(paid_fee, currency, price) => {
+                    // calculate corrected_fee in the non-native currency
+                    let converted_corrected_fee = price.checked_mul_int(corrected_fee).ok_or(TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+                    let converted_fee = paid_fee.saturating_sub(converted_corrected_fee);
+                    MC::transfer(currency.into(), &fee_receiver, who, converted_fee.into())
```

Comment:
```suggestion
                    let refund = paid_fee.saturating_sub(converted_corrected_fee);
                    MC::transfer(currency.into(), &fee_receiver, who, refund.into())
```
nitpick: should be refund